File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charalist/CharaListAdapter.java
package com.github.malitsplus.shizurunotes.ui.charalist;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.github.malitsplus.shizurunotes.R;
import com.github.malitsplus.shizurunotes.common.JsonUtils;
import com.github.malitsplus.shizurunotes.data.Chara;
import com.github.malitsplus.shizurunotes.databinding.ListItemCharaBinding;
import com.github.malitsplus.shizurunotes.ui.charadetails.CharaDetailsViewModel;

import java.util.ArrayList;
import java.util.List;

public class CharaListAdapter extends RecyclerView.Adapter<CharaListAdapter.CharaListViewHolder> {

    private Context mContext;
    private List<Chara> charaList;

    public CharaListAdapter(Context context){
        mContext = context;
        charaList = new ArrayList<>();
    }


    @Override
    @NonNull
    public CharaListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ListItemCharaBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
                R.layout.list_item_chara, parent, false);

        //向每个item设置点击监听事件
        binding.setClickListener((v) -> {
            NavDirections action = CharaListFragmentDirections.actionNavCharaToNavCharaDetails(
                    binding.getChara().unitId);
            Navigation.findNavController(v).navigate(action);
        });

        return new CharaListViewHolder(binding);

    }

    //填充每个item的视图
    @Override
    public void onBindViewHolder(CharaListViewHolder holder, int position){
        holder.getBinding().setChara(charaList.get(position));
        holder.getBinding().executePendingBindings();
    }

    @Override
    public int getItemCount(){
        return charaList.size();
    }


    public void update(List<Chara> charaList){
        this.charaList = charaList;
        notifyDataSetChanged();
    }

    //每一个item的视图控制者类（holder）
    public static class CharaListViewHolder extends RecyclerView.ViewHolder{

        private ListItemCharaBinding binding;
        CharaListViewHolder(ListItemCharaBinding binding){
            super(binding.getRoot());
            this.binding = binding;
        }

        ListItemCharaBinding getBinding(){
            return binding;
        }

    }

}

File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charalist/CharaListAdapter.kt
package com.github.malitsplus.shizurunotes.ui.charalist

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.navigation.findNavController
import androidx.navigation.fragment.FragmentNavigator
import androidx.navigation.fragment.FragmentNavigatorExtras
import androidx.recyclerview.widget.RecyclerView
import com.github.malitsplus.shizurunotes.R
import com.github.malitsplus.shizurunotes.data.Chara
import com.github.malitsplus.shizurunotes.databinding.ListItemCharaBinding
import com.github.malitsplus.shizurunotes.ui.BottomNaviFragmentDirections
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelChara
import java.util.*

class CharaListAdapter(
    private val sharedViewModelChara: SharedViewModelChara
) :
    RecyclerView.Adapter<CharaListAdapter.CharaListViewHolder>() {
    private var charaList: List<Chara> = ArrayList()
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CharaListViewHolder {
        val binding = DataBindingUtil.inflate<ListItemCharaBinding>(
            LayoutInflater.from(parent.context),
            R.layout.list_item_chara, parent, false
        )

        //向每个item设置点击监听事件
        binding.clickListener = View.OnClickListener { v: View ->
            sharedViewModelChara.mSetSelectedChara(binding.chara)
            sharedViewModelChara.backFlag = false
            var extras: FragmentNavigator.Extras
            binding.apply {
                chara?.charaId.also {
                    extras = FragmentNavigatorExtras(
                        itemChara to "transItem_$it"
                    )
                }
            }

            val action = BottomNaviFragmentDirections
                .actionNavBottomNavigationToNavCharaDetails()
                .setCharaId(binding.chara!!.charaId)

            v.findNavController().navigate(action, extras)
        }
        return CharaListViewHolder(binding)
    }


    //填充每个item的视图
    override fun onBindViewHolder(
        holder: CharaListViewHolder,
        position: Int
    ) {
        holder.binding.apply {
            chara = charaList[position]
            itemChara.transitionName = "transItem_${charaList[position].charaId}"
        }
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return charaList.size
    }

    fun update(charaList: List<Chara>) {
        this.charaList = charaList
        notifyDataSetChanged()

    }


    class CharaListViewHolder internal constructor(
        val binding: ListItemCharaBinding
    ) : RecyclerView.ViewHolder(binding.root)
}